closure_captures.md

```diff
@@ -0,0 +1,319 @@
+
+Once that is done, we walk the body again to look for
+- [`Path`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.Path.html) that represents a variable use (`Res::Local`).
+    Path isn't something like `a.b` but just represents some variable use.
+    If the variable isn't a local variable we add it to the list of upvars and map it to the Path span.
+- `Expr` that represents a Closure, i.e. handle the case of nested closures. We can query `upvars_mentioned` for the nested closure and include all the upvars returned, as part of captures for the current (enclosing) closure.
+
+
+#### [`TypeckTables::closure_captures`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TypeckTables.html#structfield.closure_captures)
+
+Generated in [`librustc_typeck`](https://github.com/rust-lang/rust/blob/9bdd2db3a60176012f4dc240eea02d615cc60061/src/librustc_typeck/check/upvar.rs#L115).
+
+This map for a given closure is very similar to upvars_mentioned. Instead of storing span we store the pair of `(closure_def_id, root_var_hir_id)`.
+
+We iterate over `upvars_mentioned` for a given `closure_def_id` to get captured root variable HirId and generate an UpvarId for them and build the map.
+
+#### [`TypeckTables::upvar_capture_map`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.TypeckTables.html#structfield.upvar_capture_map)
+
+Generated in [`librustc_typeck`](https://github.com/rust-lang/rust/blob/9bdd2db3a60176012f4dc240eea02d615cc60061/src/librustc_typeck/check/upvar.rs#L115).
+
+We implement another `intravist::Vistor` to analyse closures within a scope.
+When we see a closure within the body of the scope, we walk the body of the closure, when we are done walking the closure's body we start the analysis on the closure.
+(We need to walk first, to handle nested closures.)
+
+
+To begin the analysis we use hints like `mut` for c5, `move` in c6, to guess the initial borrow kind of each capture.
+Once the guesses are set, we create an ExprUseVistior that visits each expression within the closure body.
+ExprUseVisitor calls into MemCategorization which returns the access information about the expression as a `Place`.
+
+ExprUseVisitor delegates the work of marking a Place as consume/borrow to a callback. In our case the `Delegate` trait is implemented
+by [`InferBorrowKind`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_typeck/check/upvar/struct.InferBorrowKind.html).
+Everytime the delegate is called, it adjusts the current  borrow kind of the captured variable.
+
+It's possible that while processing the expressions in ExprUseVisitor on the closure body we see another closure (i.e. nested within the enclosing closure).
+Since we walked the body of the closure before we started the analysis, we have already proccessed all nested closures.
+We will iterate over the captured variables of this nested closure and update borrow kind for all of them for the enclosing closure.
```

Comment:
Right -- the closure expression is treated as creating a struct that captures the upvars.

Reply:
Here we actually just read the typeck tables for the closure, but yes we treat as a entity we have processed and can just read the associated information.

Reply:
https://github.com/rust-lang/rust/blob/9bdd2db3a60176012f4dc240eea02d615cc60061/src/librustc_typeck/expr_use_visitor.rs#L543

Reply:
We don't yet but we will be using the closure_captures/ the new construct here, because for the enclosed closure we have already processed capture information.